servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyPipelinedConnection.java

```diff
@@ -0,0 +1,628 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.CompletableSource;
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.Completable;
+import io.servicetalk.concurrent.api.CompletableOperator;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.PublisherOperator;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.transport.api.ExecutionContext;
+import io.servicetalk.transport.netty.internal.FlushStrategy;
+import io.servicetalk.transport.netty.internal.NettyConnection;
+import io.servicetalk.transport.netty.internal.NettyConnectionContext;
+import io.servicetalk.transport.netty.internal.WriteDemandEstimator;
+
+import io.netty.channel.Channel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.SocketAddress;
+import java.net.SocketOption;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLSession;
+
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+/**
+ * Contract for using a {@link NettyConnection} to make pipelined requests, typically for a client.
+ * <p>
+ * Pipelining allows to have concurrent requests processed on the server but still deliver responses in order.
+ * This eliminates the need for request-response correlation, at the cost of head-of-line blocking.
+ * @param <Req> Type of requests sent on this connection.
+ * @param <Resp> Type of responses read from this connection.
+ */
+final class NettyPipelinedConnection<Req, Resp> implements NettyConnectionContext {
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyPipelinedConnection.class);
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<NettyPipelinedConnection, Node>
```

Comment:
+1... I would keep it simple if possible. At first look this looks like over-engineered imho.

Reply:
I don't believe the complexity in the PR is substantially increased/decreased from what was here previously. There are two high level components of this PR (both of which existed previously):

A. relationship between read/write operations
B. Mpsc queue that only consumes and executes a single element at a time, and async completion of the current events triggers a pop of the current element and execution of the next element

(A) This PR modifies the previous control flow in [DefaultNettyPipelinedConnection](https://github.com/apple/servicetalk/blob/master/servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/DefaultNettyPipelinedConnection.java) to allow for full duplex, where previously the read could only start after the write completely finished.

(B) This PR replaces [SequentialTaskQueue](https://github.com/apple/servicetalk/blob/master/servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/SequentialTaskQueue.java) which attempted todo the same thing (Mpsc Single Runnable Active, async completion starts the next Runnable). However the previous queue relied upon `ThreadLocals` and may not make progress if the completion thread is different than the thread that started the task. This PR fixes a potential correctness issue.

The interesting characteristics of the queue are:
- offer needs to know when the queue was previously empty -> the first element needs to be executed/run
- pop needs to be correlated to the current active task (e.g. only a task that is executed/run should trigger pop, or else sequencing is off) and it also needs to execute the next task (if one exists)

this queue implementation achieves those characteristics with atomic operations to update linked nodes, and a tail pointer with trailing atomic updates. It could be done in other ways (e.g. `synchronized`) but using existing `MPSC` queues won't give us the trigger points to know when "this offer was the first element" and "this specific element needs to be popped, and I need the next element in order relative to the related read/write".

Reply:
custom queue implementation has since been removed ... thanks for the feedback!